### DISCUSSION TITLE
Q&A for beginners

### DISCUSSION THREAD

**Discussion by bipul-mohanto** [StereoKit/StereoKit#372] [Q&A] [answered]
Hi!

I found this wonderful wrapper library. As a noob, I have a few queries:

1. It has been said Stereokit allows **VR applications with C# and OpenXR!**, does it mean the library will only works on C Sharp projects? Not C++ Projects?
2. I am developing my own rendering engine for VR in C++, target device Vive HMD. I know Stereokit works on SteamVR, but will it be appropriate for the C++ project?
3. Is there eye tracking feature available that steamvr allows?

**Comment by maluoi** [comment, marked as answer]
1. StereoKit's core is a C API that contains pretty much all of the functionality exposed by the C# package! There are a number of people that do indeed consume SK from C++, and a few bindings to other languages as well! The big thing to note is that the docs are exclusively for the C# api, you can find a few C++ samples and a small C++ template in the repo, but I generally consider this to be a route for more advanced devs that don't mind poking around a bit.

2. StereoKit works through OpenXR, which means it does work well with Steam's runtime! And with C++.

3. StereoKit does implement the OpenXR extension for eye tracking, see Input.Eyes in the docs! I've only tested this on HoloLens though, but it should work fine in Steam if they've added support for that extension!

**Reply by bipul-mohanto** [reply]
Thank you very much @maluoi for your reply. If I may ask a consecutive question:

* The only C++ sample code is `StereoKitCTest` under `Examples`, right?
* The rendering part is done with **only DX11**? No DX12/ OpenGL/ any other APIs?
* Is there any limitation using any other API for rendering, e.g., NVIDIA's OptiX?

**Reply by maluoi** [reply]
Yeah absolutely! :)

1. [StereoKitCTest](https://github.com/maluoi/StereoKit/tree/master/Examples/StereoKitCTest) is the largest collection of C++ example code, yes! The C# code also generally translates pretty directly to the C API, so if for example you see in C#: `modelInst.Draw(...)`, this translates to C++ as: `model_draw(modelInst, ...)`. So don't completely discount the C# documentation! It may also be worth checking the [`/develop`](https://github.com/maluoi/StereoKit/tree/develop/Examples/StereoKitCTest) branch too for new stuff :)
The [native template](https://github.com/maluoi/StereoKit/tree/develop/Templates/SKNative) for C++ is designed a bit more for cross-platform work, so it uses a cmake project rather than a VS solution. But if you're a VS person, you can always use cmake to generate a VS project!
2. Rendering is done via DirectX11, OpenGL, OpenGLES, or WebGL depending on the platform. There's a [separate repository](https://github.com/maluoi/sk_gpu) for the graphics abstraction if you're interested in that!
3. Adding another graphics API to StereoKit would likely be a big chunk of work! If you can properly hide it behind `sk_gpu.h`'s API, then it'll mostly just work in StereoKit. SK is currently set up to support only one rendering API at a time, at compile-time. I don't know anything about OptiX though, so I can't speak to that.
